MoreLinq/Experimental/Await.cs

```diff
@@ -531,149 +610,82 @@ IEnumerable<TResult> _(int maxConcurrency, TaskScheduler scheduler, bool ordered
-                    // Task.WaitAny is synchronous and blocking but allows the
-                    // waiting to be cancelled via a CancellationToken.
-                    // Task.WhenAny can be awaited so it is better since the
-                    // thread won't be blocked and can return to the pool.
-                    // However, it doesn't support cancellation so instead a
-                    // task is built on top of the CancellationToken that
-                    // completes when the CancellationToken trips.
-                    //
-                    // Also, Task.WhenAny returns the task (Task) object that
-                    // completed but task objects may not be unique due to
-                    // caching, e.g.:
-                    //
-                    //     async Task<bool> Foo() => true;
-                    //     async Task<bool> Bar() => true;
-                    //     var foo = Foo();
-                    //     var bar = Bar();
-                    //     var same = foo.Equals(bar); // == true
-                    //
-                    // In this case, the task returned by Task.WhenAny will
-                    // match `foo` and `bar`:
-                    //
-                    //     var done = Task.WhenAny(foo, bar);
-                    //
-                    // Logically speaking, the uniqueness of a task does not
-                    // matter but here it does, especially when Await (the main
-                    // user of CollectAsync) needs to return results ordered.
-                    // Fortunately, we compose our own task on top of the
-                    // original that links each item with the task result and as
-                    // a consequence generate new and unique task objects.
-
-                    var completedTask = await
-                        Task.WhenAny(tasks.Select(it => (Task) it.Task).Concat(cancellationTaskSource.Task))
-                            .ConfigureAwait(continueOnCapturedContext: false);
-
-                    if (completedTask == cancellationTaskSource.Task)
+                    if (semaphore != null)
                     {
-                        // Cancellation during the wait means the enumeration
-                        // has been stopped by the user so the results of the
-                        // remaining tasks are no longer needed. Those tasks
-                        // should cancel as a result of sharing the same
-                        // cancellation token and provided that they passed it
-                        // on to any downstream asynchronous operations. Either
-                        // way, this loop is done so exit hard here.
-
-                        return;
+                        try
+                        {
+                            await semaphore.WaitAsync(cancellationToken);
+                        }
+                        catch (OperationCanceledException e) when (e.CancellationToken == cancellationToken)
```

Comment:
This `catch` block could be moved back up into the caller, and below make change the cancellation token check to `cancellationToken.ThrowIfCancellationRequested();` . This should make this function simple, while not making the caller too complicated as it is already catching exceptions.

Reply:
The `catch` in the caller is for exceptional issues in `StartAsync`, like invocation of `starter` throwing (task fails to start). The returns are normal conditions or exits that are accounted for.